### PR TITLE
feat: add error code mapping for gw/polyjuice/evmc

### PR DESCRIPTION
**Pull request by RetricSu**
some examples:

contract revert 
```sh
{
    "jsonrpc": "2.0",
    "id": 1,
    "error": {
        "code": -32000,
        "message": "execution reverted: here is a simple revert reason",
        "data": "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000001e6865726520697320612073696d706c652072657665727420726561736f6e0000",
        "extra": {
            "exit_code": "0x2",
            "message": "revert",
            "stack": "EVMC"
        }
    }
}
```

max cycles
```sh
{
    "jsonrpc": "2.0",
    "id": 1,
    "error": {
        "code": -32000,
        "message": "out of gas(vm reached max cycles)",
        "extra": {
            "exit_code": "0xff",
            "message": "vm reached max cycles",
            "stack": "GODWOKEN"
        }
    }
}
```
solidity call baseFee opcode
```sh
{
    "jsonrpc": "2.0",
    "id": 1,
    "error": {
        "code": -32000,
        "message": "undefined instruction",
        "extra": {
            "exit_code": "0x5",
            "message": "undefined instruction",
            "stack": "EVMC"
        }
    }
}
```

### CODE REVIEW

**Review by magicalne — APPROVED** [review]

File: packages/api-server/src/methods/exit-code.ts [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+    message: "gw registry error duplicate mapping",
+  },
+};
+
+export function matchExitCode(
+  exitCode: HexNumber | number,
+  mapping: ExitCodeMapping
+): ExitCode | null {
+  let code: number;
+  if (typeof exitCode == "number") {
```
Comment (classicalliu) [inline packages/api-server/src/methods/exit-code.ts:370]:
```suggestion
  if (typeof exitCode === "number") {
```

File: packages/api-server/src/methods/exit-code.ts [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+  let num = parseInt(hex, 16);
+  if (num > 127) {
+    num = num - 256;
+  }
+  return num;
+}
+
+// the exit_code in godwoken is always an i8 type.
+export function exitCodeNumberToHex(_num: number): HexNumber {
+  let num = _num;
```
Comment (classicalliu) [inline packages/api-server/src/methods/exit-code.ts:398]:
```suggestion
export function exitCodeNumberToHex(num: number): HexNumber {
```

File: packages/api-server/src/methods/gw-error.ts [hunk 3]
```diff
@@ -74,24 +91,108 @@ export function handleGwError(gwJsonRpcError: any) {
...
+      message = gwCode.message;
+    }
   }
+  return {
+    message,
+    extraMessage,
+  };
+}
+
+function unpackPolyjuiceSystemLog(logItem: LogItem): PolyjuiceSystemLog {
```
Comment (classicalliu) [inline packages/api-server/src/methods/gw-error.ts:181]: https://github.com/godwokenrises/godwoken-web3/blob/c456a58a1126c1bd0d19c0d5142d2a9904fc3422/packages/api-server/src/filter-web3-tx.ts#L266
This function already exists

Comment (RetricSu) [reply packages/api-server/src/methods/gw-error.ts]: updated

**Review by classicalliu — APPROVED** [review]

**Review by Flouse — APPROVED** [review]

**Pull request merged**